networkx/algorithms/shortest_paths/ada_star.py

```diff
@@ -0,0 +1,470 @@
+
+
+    Raises
+    ------
+    NetworkXNoPath
+        If no path exists between source and target.
+    NodeNotFound
+        If either source or target is not in G.
+
+    See Also
+    --------
+    shortest_path, astar_path, dijkstra_path
+
+    References
+    ----------
+    [1] Likhachev, Maxim, David I. Ferguson, Geoffrey J. Gordon, Anthony
+    Stentz, and Sebastian Thrun. "Anytime dynamic A*: An anytime,
+    replanning algorithm." In ICAPS, vol. 5, pp. 262-271. 2005.
+
+    Examples
+    --------
+    >>> import math
+
+    >>> G = nx.random_geometric_graph(100, 0.20, seed=896803)
+    >>> for u, v, w in G.edges(data=True):  # Euclidean distance between nodes
+    ...     w["weight"] = math.sqrt(
+    ...         (G.nodes[v]["pos"][0] - G.nodes[u]["pos"][0]) ** 2
+    ...         + (G.nodes[v]["pos"][1] - G.nodes[u]["pos"][1]) ** 2
+    ...     )
+    >>> source, target = 42, 25
+
+    >>> def heursistic(u, v):  # Euclidean distance between nodes
+    ...     return math.sqrt(
+    ...         (G.nodes[v]["pos"][0] - G.nodes[u]["pos"][0]) ** 2
+    ...         + (G.nodes[v]["pos"][1] - G.nodes[u]["pos"][1]) ** 2
+    ...     )
+
+    >>> # A* search for comparison
+    >>> path = nx.astar_path(G, source, target, heursistic)
+    >>> print("A* path: ", path)
+    A* path:  [42, 32, 19, 72, 49, 29, 31, 94, 35, 25]
+    >>> search = ada_star(source, target, G, heursistic)
+
+    >>> search.compute_or_improve_path(epsilon=2)
+    >>> path = search.extract_path()
+    >>> print("epsilon = 2 path: ", path)
+    epsilon = 2 path:  [42, 32, 24, 40, 59, 4, 66, 27, 35, 25]
+    >>> print(
+    ...     "epsilon = 2 path_weight: ", nx.path_weight(G, search.extract_path(), "weight")
+    ... )
+    epsilon = 2 path_weight:  1.4679609830956495
+
+    >>> search.compute_or_improve_path(epsilon=1.2)
+    >>> path = search.extract_path()
+    >>> print("epsilon = 1.2 path: ", path)
+    epsilon = 1.2 path:  [42, 32, 24, 12, 59, 4, 1, 27, 35, 25]
+    >>> print("epsilon = 1.2 path_weight: ", nx.path_weight(G, path, "weight"))
+    epsilon = 1.2 path_weight:  1.3335657361796027
+
+    >>> search.compute_or_improve_path(epsilon=1)
+    >>> path = search.extract_path()
+    >>> print("epsilon = 1 path: ", path)
+    epsilon = 1 path:  [42, 32, 19, 72, 49, 29, 31, 94, 35, 25]
+    >>> print("epsilon = 1 path_weight: ", nx.path_weight(G, path, "weight"))
+    epsilon = 1 path_weight:  1.29129785933092
+
+    >>> search.update_graph([[49, 97, 0]])  # set weight between 49 and 97 to 0
+    >>> search.compute_or_improve_path(epsilon=1)
+    >>> path = search.extract_path()
+    >>> print("changed epsilon = 1 path: ", path)
+    changed epsilon = 1 path:  [42, 32, 19, 72, 49, 29, 31, 94, 35, 25]
+    >>> print("changed epsilon = 1 path_weight: ", nx.path_weight(G, path, "weight"))
+    changed epsilon = 1 path_weight:  1.29129785933092
+    """
+
+    def __init__(
+        self, source, target, G, heuristic=None, weight="weight", initial_epsilon=1000
```

Comment:
This should have `G` as the first argument and then source target.  Then make the rest keyword-only by putting a * symbol in the function signature. That helps users keep from switching order by mistake and also makes their code more readable in the long run.  Some of the tests in this PR would be more readable with the keyword names entered -- so this change will require changing the tests some too.

Reply:
I have changed the order to reflect the networkx convention